### PR TITLE
Add proof annotation option

### PR DESCRIPTION
**Pull request by ajreynol**
Adds `--proof-annotate` which enables a useful stat in which we count the number of times a lemma with a given inference id appears in a final proof.

This can be used to determine which lemma schemas are the most useful for showing a problem unsat.

### CODE REVIEW

**Review by HanielB — CHANGES_REQUESTED** [review]

File: src/proof/annotation_proof_generator.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * A user-context-dependent map from formulas to a generator + annotation
+   * pair, which will be used to generate the proof of formulas if asked.
+   */
+  NodeExpMap d_exps;
+  /**
+   * A user-context-dependent map from formulas to the proof nodes we have
+   * returned in calls to getProofFor.
+   */
+  NodeProofNodeMap d_proofs;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/proof/annotation_proof_generator.h:96]: The documentation is off, I'll update.

File: src/theory/inference_id_proof_annotator.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+{
+  Node f = p->getResult();
+  NodeInferenceIdMap::iterator it = d_ids.find(f);
+  if (it != d_ids.end())
+  {
+    std::vector<Node> pfArgs;
+    pfArgs.push_back(mkInferenceIdNode(it->second));
+    std::shared_ptr<ProofNode> pa =
+        d_pnm->mkNode(PfRule::ANNOTATION, {p}, pfArgs);
+    // for now, do a double annotation to make stats accurate
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/inference_id_proof_annotator.cpp:45]: There was a technical discussion we had at some point about changing how `ProofNodeManager::updateNode` worked, but it led to pretty major complications involving cyclic proofs as I recall.

File: src/theory/inference_id_proof_annotator.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * The inference id annotator class.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__INFERENCE_ID_PROOF_ANNOTATOR_H
+#define CVC5__THEORY__INFERENCE_ID_PROOF_ANNOTATOR_H
+
+#include "context/cdhashmap.h"
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/inference_id_proof_annotator.h:21]: I'm following include-what-you-use style here.

File: src/theory/inference_id_proof_annotator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/inference_id_proof_annotator.cpp:45]: I remember now... yeah, maybe let's not touch that for now.

**Review by HanielB — APPROVED** [review]

**Pull request merged**